src/v/cloud_storage/materialized_segments.cc

```diff
@@ -0,0 +1,165 @@
+
+    co_await _gate.close();
+
+    // Do the last pass over the eviction list to stop remaining items returned
+    // from readers after the eviction loop stopped.
+    for (auto& rs : _eviction_list) {
+        co_await std::visit([](auto&& rs) { return rs->stop(); }, rs);
+    }
+}
+ss::future<> materialized_segments::start() {
+    // Fiber that consumes from _eviction_list and calls stop
+    // on items before destroying them
+    ssx::spawn_with_gate(_gate, [this] { return run_eviction_loop(); });
+
+    // Timer to invoke TTL eviction of segments
+    _stm_timer.set_callback([this] {
+        gc_stale_materialized_segments(false);
+        _stm_timer.rearm(_stm_jitter());
+    });
+    _stm_timer.rearm(_stm_jitter());
+
+    return ss::now();
+}
+
+void materialized_segments::evict_reader(
+  std::unique_ptr<remote_segment_batch_reader> reader) {
+    _eviction_list.push_back(std::move(reader));
+    _cvar.signal();
+}
+void materialized_segments::evict_segment(
+  ss::lw_shared_ptr<remote_segment> segment) {
+    _eviction_list.push_back(std::move(segment));
+    _cvar.signal();
+}
+
+ss::future<> materialized_segments::run_eviction_loop() {
+    // Evict readers asynchronously
+    while (true) {
+        co_await _cvar.wait([this] { return !_eviction_list.empty(); });
+        auto tmp_list = std::exchange(_eviction_list, {});
+        for (auto& rs : tmp_list) {
+            co_await std::visit([](auto&& rs) { return rs->stop(); }, rs);
+        }
+    }
+}
+
+void materialized_segments::register_segment(materialized_segment_state& s) {
+    _materialized.push_back(s);
+}
+
+void materialized_segments::maybe_trim() {
+    // TODO: implement trimming to bound resource consumption
+}
+
+void materialized_segments::gc_stale_materialized_segments(
+  bool force_collection) {
+    // TODO: augment the log messages in these functions with ntp
+
+    // The remote_segment instances are materialized on demand. They are
+    // collected after some period of inactivity.
+    // To prevent high memory consumption in some corner cases the
+    // materialization of the new remote_segment triggers GC. The idea is
+    // that remote_partition should have only one remote_segment in materialized
+    // state when it's constantly in use and zero if not in use.
+    vlog(
+      cst_log.debug,
+      "collecting stale materialized segments, {} segments materialized",
+      _materialized.size());
+
+    auto now = ss::lowres_clock::now();
+    auto max_idle = force_collection ? 0ms : stm_max_idle_time;
+
+    // These pointers are safe because there are no scheduling points
+    // between here and the ultimate eviction at end of function.
+    std::vector<std::pair<materialized_segment_state*, kafka::offset>>
+      to_offload;
+
+    for (auto& st : _materialized) {
+        auto deadline = st.atime + max_idle;
+        if (now >= deadline && !st.segment->download_in_progress()) {
+            if (st.segment.owned()) {
+                vlog(
```

Comment:
the log line shuld have an ntp because many partitions could use the same GC loop

Reply:
Right, I'll add it in the followup PR.